vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,283 @@
+    Unknown,
+}
+
+impl From<u64> for EpollDispatch {
+    fn from(v: u64) -> Self {
+        use EpollDispatch::*;
+        match v {
+            0 => File,
+            1 => Kill,
+            _ => Unknown,
+        }
+    }
+}
+
+pub struct SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    serial: Arc<Mutex<Serial>>,
+    #[cfg(target_arch = "aarch64")]
+    serial: Arc<Mutex<Pl011>>,
+    epoll_file: File,
+    in_file: File,
+    kill_evt: EventFd,
+}
+
+impl SerialManager {
+    #[cfg(target_arch = "x86_64")]
+    pub fn new(
+        serial: Arc<Mutex<Serial>>,
+        pty_file: Option<File>,
+        mode: ConsoleOutputMode,
+        kill_evt: EventFd,
+    ) -> Result<Option<Self>> {
+        let epoll_fd = epoll::create(true).map_err(Error::Epoll)?;
+
+        let in_file = match mode {
+            ConsoleOutputMode::Pty => {
+                let in_file_clone = pty_file.unwrap().try_clone().map_err(Error::FileClone)?;
+                epoll::ctl(
+                    epoll_fd,
+                    epoll::ControlOptions::EPOLL_CTL_ADD,
+                    in_file_clone.as_raw_fd(),
+                    epoll::Event::new(epoll::Events::EPOLLIN, EpollDispatch::File as u64),
+                )
+                .map_err(Error::Epoll)?;
+                Some(in_file_clone)
+            }
+            ConsoleOutputMode::Tty => {
+                // If running on an interactive TTY then accept input
+                if unsafe { libc::isatty(libc::STDIN_FILENO) == 1 } {
+                    let stdin_clone = unsafe { File::from_raw_fd(libc::dup(libc::STDIN_FILENO)) };
+                    let ret = unsafe {
+                        let mut flags = libc::fcntl(stdin_clone.as_raw_fd(), libc::F_GETFL);
+                        flags |= libc::O_NONBLOCK;
+                        libc::fcntl(stdin_clone.as_raw_fd(), libc::F_SETFL, flags)
+                    };
+
+                    if ret < 0 {
+                        return Err(Error::SetNonBlocking(std::io::Error::last_os_error()));
+                    }
+
+                    epoll::ctl(
+                        epoll_fd,
+                        epoll::ControlOptions::EPOLL_CTL_ADD,
+                        stdin_clone.as_raw_fd(),
+                        epoll::Event::new(epoll::Events::EPOLLIN, EpollDispatch::File as u64),
+                    )
+                    .map_err(Error::Epoll)?;
+                    Some(stdin_clone)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        };
+
+        // Use 'File' to enforce closing on 'epoll_fd'
+        let epoll_file = unsafe { File::from_raw_fd(epoll_fd) };
+
+        if let Some(in_file) = in_file {
+            Ok(Some(SerialManager {
+                serial,
+                epoll_file,
+                in_file,
+                kill_evt,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    #[cfg(target_arch = "aarch64")]
```

Comment:
These constructors are basically identical. Can't you put the attribute on the function parameter instead?

Reply:
(I pushed a change that did that.)

Reply:
Thanks!